### ISSUE TITLE
Option to disable storing any history

### ISSUE THREAD

**Issue by infozsoda** [grote/Transportr#263]
It would be nice to have a "private querying" toggle in the settings, to prevent any history from being stored. Like "private browsing" in Firefox.

**Comment by grote** [comment]
Thanks for the suggestion. That's a good idea!

The first step for implementing the feature is to think about how it could be integrated into the current Ui in a way that doesn't make it more complicated to use and still enables users who are interested in this feature to discover it.

**Comment by jobot0** [comment]
Hello, 

I've discovered the application pretty recently and I really like it. I was looking to contribute and I'm trying to implement this feature. I'm not sure if I'm understand it correctly, The feature should provide like Firefox an anonymous account during all the "session" with all the preferences empty or something to hide the trips searched during the feature is enabled? For both case I think a new Preference ("Private mode", "Incognito Trips") could let the more advanced users find it without complicates the current one?

Thanks again for this nice application!

**Comment by grote** [comment]
Thanks @jobot0 for you willingness to contribute! :)

The easiest way to implement this would indeed be just another setting that is checked before data is stored to the database (in all forms).

However, this way of implementing it has some downsides:
* difficult to discover that the feature exists
* no indication that the feature is active (this might look like a bug to people who enabled that by mistake)

If there's a better way to integrate this in the existing UI, I would certainly prefer that.

**Comment by jobot0** [comment]
I'm currently at work and can't provide any picture but we could directly add an incognito icon in the toolbar and change the text in the button to do the search when the mode is enabled? So it's directly available to any user and the indication is pretty clear. The button view will be a bit overloaded though.

**Comment by grote** [comment]
The toolbar has limited space for additional icons. They could go to the overflow menu though. However, the incognito mode should be universal since it would not only apply when searching for directions, but also for departures and nearby stations.

Maybe there could be a toggle in the navdrawer and only shows up once incognito mode has been activated in the settings?

**Comment by Philip-W** [comment]
Hey! Is anyone still working on this? I'm also interested in contributing and thinking this might be a decent starting point.

**Comment by grote** [comment]
For those interested in more privacy options, other possible features are support for [panic button](https://dev.guardianproject.info/projects/panic/wiki) apps and Tor support (#265).

**Comment by jobot0** [comment]
Yes you're right there is more room for this in the navdrawer. @Philip-W I'm still working on it but you should go for it if you want too :)

**Comment by Philip-W** [comment]
Hey what are your thoughts on a change to the color scheme when browsing privately? Much like Firefox does!? I think I've got the actual setting sorted (it stops entries as required) it's just missing visuals as suggested.

**Comment by grote** [comment]
Great to hear that there is progress.

Changing colors in private mode is possible. It depends on the details.

> Or perhaps a less drastic change to the visual could be to add card views in the recent search fragment informing the user their searches aren't being recorded?

That doesn't sound good to be honest and again, there are other search forms as well (departures, nearby stations) where this notice wouldn't be visible.

**Comment by jobot0** [comment]
I need to try different things to display incognito icon/warning on the navdrawer and I'll publish some screenshots to show you how it renders. 

Changing the theme could be a nice solution but the user can already change it (between light/dark) so it may be hard to find something incognito compare to the one the user already chose.

**Comment by Philip-W** [comment]
You both raise good points! I was unaware the user had the ability to change the theme, I think if you have a decent way to display it on the UI I shall leave this task to you!

**Comment by grote** [comment]
> Changing the theme could be a nice solution but the user can already change it (between light/dark) so it may be hard to find something incognito compare to the one the user has already chosen.

I also think that changing the entire theme won't work. I was more thinking about changing the toolbar (action bar) color. Alternatively, there could maybe be a privacy mode icon in all search buttons when privacy mode is activated.

However, as written above there's also a Tor ticket, so we need to be careful that privacy mode doesn't give the impression that people are completely anonymous with their searches.

**Comment by craftyguy** [comment]
I'm of the opinion that if 1) there were a "privacy" section in Settings where this option was listed and 2) the "getting started" screen when the app is first installed and launched mentioned/linked to the privacy section in Settings, that most folks who wanted to use it could find it. 


If you want to indicate that it's active, add a simple icon to the action bar to indicate that (I could help find one), or don't since the option would have to be explicitly enabled and you can't be expect to babysit users at all times.

**Comment by jobot0** [comment]
I'm trying a new layout, I'm just adding a incognito cardview info on top of the current search form (Directions, Departures and Nearby Stations). Here a screenshot.
![screenshot](https://cloud.githubusercontent.com/assets/20325318/22184737/c1a24ce4-e0d8-11e6-8a24-4399db351c86.png)

**Comment by grote** [comment]
Looks nice!

> I could also put a dialog when the user is doing a long click on the card to explain what the incognito mode is.

Users who need an explanation don't know what a long-click is and where to do it. A short click would do it, but I expect the explanation already to be in the settings when the user activated the mode, so that no further explanation is required.

Btw. I started work on Transportr 2 in a local branch which will change a lot of things. The sooner a PR for this gets merged, the better.

**Comment by sebalis** [comment]
Apart from not having an individual trip enter my history, I also think we need to be able to clear it. See #343.

**Comment by sebalis** [comment]
Well, if we will finally get some progress on #74 … I am still very interested! :-) And it is still the case that the inability to weed out the history is a showstopper for me. I do want to support Transportr, if not by contributing then at least by word of mouth – as you know, Torsten ;-)

**Comment by grote** [comment]
As a workaround, you can still clean the app's data in Android's app settings for now.

**Comment by sebalis** [comment]
When we spoke about this in person (before version 2) you suggested switching providers. Both of these are far from perfect. Deleting the data forgets my choice of provider and home/work locations. I very strongly hope that this does not indicate a disregard for the importance of the issue – I would not expect that from you. :-)

**Comment by grote** [comment]
I just personally don't have a use-case for this that would motivate me enough to put ~20h of my free time into the implementation.

And yes, if your use-case is just to show the app to somebody else without letting them see your favorites, you can switch to a different transport network before showing it.

**Comment by sebalis** [comment]
This is and remains a showstopper for me both for privacy and usability reasons. I do want to use this app but the impossibility of weeding out the history really puts me off. I find myself using a browser to [DB](https://reiseauskunft.bahn.de/) instead.

**Comment by grote** [comment]
> for being able to find those locations in the history that I might want to reuse, rather than having them disappear in a noise of any stop I ever entered (be it for a one-time journey or merely for a trial of some sort).

Locations should be sorted by how often you use them. If you use a stop frequently, it should be at the top. If you just used one once, it should be at the bottom.

> I realise I can’t coerce you into doing work you don’t have the capacity or intrinsic motivation to do. 

You could pay me ;)

> But please do acknowledge that this is potentially important from several perspectives (you do know about the importance of privacy issues).

Sure, I get that this is important for people and that their might be valid use-cases. They are just not motivating enough for *me* to spend unpaid free time on it.

> the impossibility of weeding out the history really puts me off

But this is a different issue. It is #74

**Comment by sebalis** [comment]
> Locations should be sorted by how often you use them

Be aware of the small but important difference between “might want to reuse” and “have actually resued often enough to get the history entry pushed to the top” :-)

> But this is a different issue. It is #74

That is true, this is why I started the revival of this discussion with the word “Well, if we will finally get some progress on #74”, hesitantly accepting the closure of this one :-)

**Comment by sebalis** [comment]
Added the 'unresolved' label although it seems like #74 (unclosed) should be given priority as it would offer more flexibility (i.e. delete only some unwanted entries).
